Changes in module 'nf-core/blast/blastn'
--- modules/nf-core/blast/blastn/main.nf
+++ modules/nf-core/blast/blastn/main.nf
@@ -19,10 +19,12 @@
     task.ext.when == null || task.ext.when
 
     script:
-    def args = task.ext.args ?: ''
-    def prefix = task.ext.prefix ?: "${meta.id}"
+    def args        = task.ext.args ?: ''
+    def args2       = task.ext.args2 ?: '' // for ascc this is " | head -n1"
+    def prefix      = task.ext.prefix ?: "${meta.id}"
+    def dbprefix    = task.ext.dbprefix ?: ".nin" // for ascc this is ".[0-9]*\\.nin" as the database has many nt.{interger}.nin
     """
-    DB=`find -L ./ -name "*.nin" | sed 's/\\.nin\$//'`
+    DB=`find -L ./ -name "*.nin" | sed 's/\\${dbprefix}\$//'${args2}`
     blastn \\
         -num_threads $task.cpus \\
         -db \$DB \\

************************************************************
