Changes in module 'nf-core/diamond/blastx'
--- modules/nf-core/diamond/blastx/main.nf
+++ modules/nf-core/diamond/blastx/main.nf
@@ -2,10 +2,10 @@
     tag "$meta.id"
     label 'process_medium'
 
-    conda "bioconda::diamond=2.0.15"
+    conda "bioconda::diamond=2.1.8"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/diamond:2.0.15--hb97b32f_0' :
-        'biocontainers/diamond:2.0.15--hb97b32f_0' }"
+        'https://depot.galaxyproject.org/singularity/diamond:2.1.8--h43eeafb_0' :
+        'biocontainers/diamond:2.1.8--h43eeafb_0' }"
 
     input:
     tuple val(meta), path(fasta)

************************************************************
